drivers/espi/espi_mchp_xec_v2.c

```diff
@@ -1561,3 +1561,282 @@ static int espi_xec_init(const struct device *dev)
+		ret = espi_read_flash(espi_dev, &pkt);
+		if (ret) {
+			LOG_ERR("espi flash read failed: %d", ret);
+			return ret;
+		}
+
+		mem += MAX_FLASH_REQUEST;
+		flash_addr += MAX_FLASH_REQUEST;
+		remaining -= MAX_FLASH_REQUEST;
+		transactions--;
+	}
+	if (remaining) {
+		LOG_DBG("%s: Remaining bytes = %d, offset = 0x%x", __func__, remaining, flash_addr);
+		pkt.buf = mem;
+		pkt.flash_addr = flash_addr;
+		pkt.len = remaining;
+
+		ret = espi_read_flash(espi_dev, &pkt);
+	}
+
+	k_sem_give(&dev_data->mutex);
+
+	if (ret) {
+		LOG_ERR("espi flash read failed: %d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int flash_espi_write(const struct device *dev, off_t offset,
+			    const void *data, size_t len)
+{
+	uint32_t flash_addr = offset;
+	uint16_t transactions = len / MAX_FLASH_REQUEST;
+	uint16_t remaining = len;
+	uint8_t *mem = (uint8_t *)data;
+	struct espi_flash_packet pkt;
+	struct flash_espi_data *dev_data = dev->data;
+	int ret = 0;
+
+	if (!flash_espi_valid_range(offset, len, true)) {
+		return -EINVAL;
+	}
+
+	if (len == 0U) {
+		return 0;
+	}
+
+	/*
+	 * create FLASH channel packets to write
+	 */
+	k_sem_take(&dev_data->mutex, K_FOREVER);
+	if ((flash_addr & 0x3F) && (ROUND_DOWN(offset, 4096) != ROUND_DOWN((offset + len), 4096))) {
+		pkt.buf = mem;
+		pkt.flash_addr = flash_addr;
+		pkt.len = ROUND_UP(flash_addr, 64) - flash_addr;
+
+		ret = espi_write_flash(espi_dev, &pkt);
+		if (ret) {
+			LOG_ERR("espi flash_read < 64B failed: %d", ret);
+			return ret;
+		}
+		mem += pkt.len;
+		flash_addr += pkt.len;
+		remaining -= pkt.len;
+	}
+
+	while (transactions > 1) {
+		pkt.buf = mem;
+		pkt.flash_addr = flash_addr;
+		pkt.len = MAX_FLASH_REQUEST;
+
+		ret = espi_write_flash(espi_dev, &pkt);
+		if (ret) {
+			LOG_ERR("espi flash read failed: %d", ret);
+			return ret;
+		}
+
+		mem += MAX_FLASH_REQUEST;
+		flash_addr += MAX_FLASH_REQUEST;
+		remaining -= MAX_FLASH_REQUEST;
+		transactions--;
+	}
+	if (remaining) {
+		pkt.buf = mem;
+		pkt.flash_addr = flash_addr;
+		pkt.len = remaining;
+
+		ret = espi_write_flash(espi_dev, &pkt);
+	}
+
+	k_sem_give(&dev_data->mutex);
+
+	if (ret) {
+		LOG_ERR("espi flash write failed: %d", ret);
+		return ret;
+	}
+
+	return ret;
+}
+
+static int flash_espi_erase(const struct device *dev, off_t offset, size_t size)
+{
+	uint32_t flash_addr = offset;
+	uint16_t transactions = size / 4096;
+	struct espi_flash_packet pkt;
+	struct flash_espi_data *dev_data = dev->data;
+	int ret = 0;
+
+	if (size == 0U) {
+		return 0;
+	}
+
+	if (!flash_espi_valid_range(offset, size, false)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * create FLASH channel packets to erase
+	 */
+	k_sem_take(&dev_data->mutex, K_FOREVER);
+	while (transactions > 1) {
+		pkt.flash_addr = flash_addr;
+		pkt.len = 4096;
+
+		ret = espi_flash_erase(espi_dev, &pkt);
+		if (ret) {
+			LOG_ERR("espi flash read failed: %d", ret);
+			return ret;
+		}
+
+		flash_addr += 4096;
+		transactions--;
+	}
+	pkt.flash_addr = flash_addr;
+	pkt.len = 4096;
```

Comment:
Magic number. Consider using value of Flash Block Erase Size from eSPI controller set during eSPI flash channel negotiation.